src/Akka.Persistence.MongoDb/Journal/MongoDbJournal.cs

```diff
@@ -410,95 +433,180 @@ protected override bool ReceivePluginInternal(object message)
+
+            var seqNoFilter = builder.Empty;
+            if (fromSequenceNr > 0)
+                seqNoFilter &= builder.Gt(x => x.Ordering, new BsonTimestamp(fromSequenceNr));
+            if (toSequenceNr != long.MaxValue)
+                seqNoFilter &= builder.Lte(x => x.Ordering, new BsonTimestamp(toSequenceNr));
+
+
+            // Need to know what the highest seqNo of this query will be
+            // and return that as part of the RecoverySuccess message
+            var maxSeqNoEntry = await _journalCollection.Value.Find(seqNoFilter)
+                .SortByDescending(x => x.Ordering)
+                .Limit(1)
+                .SingleOrDefaultAsync();
+
+            if (maxSeqNoEntry == null)
+                return 0L; // recovered nothing
+
+            var maxOrderingId = maxSeqNoEntry.Ordering.Value;
+            var toSeqNo = Math.Min(toSequenceNr, maxOrderingId);
+
+            var readFilter = builder.Empty;
+            if (fromSequenceNr > 0)
+                readFilter &= builder.Gt(x => x.Ordering, new BsonTimestamp(fromSequenceNr));
+            if (toSequenceNr != long.MaxValue)
+                readFilter &= builder.Lte(x => x.Ordering, new BsonTimestamp(toSeqNo));
+            var sort = Builders<JournalEntry>.Sort.Ascending(x => x.Ordering);
+
+            await _journalCollection.Value.Find(readFilter)
+                .Sort(sort)
+                .Limit(limitValue)
+                .ForEachAsync(entry =>
+                {
+                    var persistent = ToPersistenceRepresentation(entry, ActorRefs.NoSender);
+                    foreach (var adapted in AdaptFromJournal(persistent))
+                    {
+                        replay.ReplyTo.Tell(new ReplayedEvent(adapted, entry.Ordering.Value), ActorRefs.NoSender);
+                    }
+                });
+            return maxOrderingId;
         }
 
         private void AddTagSubscriber(IActorRef subscriber, string tag)
         {
             if (!_tagSubscribers.TryGetValue(tag, out var subscriptions))
             {
-                subscriptions = new HashSet<IActorRef>();
-                _tagSubscribers.Add(tag, subscriptions);
+                _tagSubscribers = _tagSubscribers.Add(tag, ImmutableHashSet.Create(subscriber));
+            }
+            else
+            {
+                _tagSubscribers = _tagSubscribers.SetItem(tag, subscriptions.Add(subscriber));
             }
-
-            subscriptions.Add(subscriber);
         }
 
         private IEnumerable<string> GetAllPersistenceIds()
         {
             return _journalCollection.Value.AsQueryable()
-                .Select(je => je.PersistenceId)
-                .Distinct()
-                .ToList();
+                 .Select(je => je.PersistenceId)
+                 .Distinct()
+                 .ToList();
         }
 
+        private async Task<long> GetHighestOrdering()
+        {
+            return await Task.Run(() =>
```

Comment:
Same comment as below - just use the `MaxAsync` call

Reply:
Also, this query can be rewritten in a much less expensive way:

https://github.com/akkadotnet/Akka.Persistence.MongoDB/blob/38a0c4996f610e4e03e32dcccea9ed08aa534a6a/src/Akka.Persistence.MongoDb/Journal/MongoDbJournal.cs#L82-L85

We have a sorted index built on top of the `Ordering` column. Just select the maximum value from it. Don't do an entire table scan. That turns this into a O(1) query rather than an O(n) query.